plugins/mac-capture/mac-sck-video-capture.m

```diff
@@ -142,8 +142,14 @@ static bool init_screen_stream(struct screen_capture *sc)
                     }
                 }
             } else {
-                target_window = [sc->shareable_content.windows objectAtIndex:0];
-                sc->window = target_window.windowID;
+                target_window = nil;
```

Comment:
```suggestion
            SCWindow *target_window = nil;
            if (sc->window != 0) {
                for (SCWindow *window in sc->shareable_content.windows) {
                    if (window.windowID == sc->window) {
                        target_window = window;
                        break;
                     }
                 }
             }
             
             // sc->window was actually 0 or the window ID could not be found
             if (!target_window) {
                 // This assumes that there should be more than a single window
                 // in the array
                 target_window = [sc->shareable_content.windows objectAtIndex:0];
                 sc->window = target_window.windowID;
             }             
```

Reply:
Doesn't this introduce a lot more scenarios where we are capturing windows that the user didn't pick? Isn't it preferable to return early instead of capturing the first window?

I'm unclear if id=0 is even legitimate. I have not ever seen it in `SCShareableContent`'s window list, though I haven't done super thorough testing to see if it ever appears. I know that the system starts incrementing windowIDs on boot, but I'm not sure if it ever wraps around. There seem to be existing checks in this code ensuring that the window ID is nonzero as well?

This suggested code picks `[sc->shareable_content.windows objectAtIndex:0]` without checking if its windowID is actually 0; isn't that incorrect?

Reply:
I think you're hitting an interesting point:

* Logically it should not be an issue to fall back on the very first item in the list, because you're configuring the source and you haven't committed the change yet - you're just setting it up. In that case there should not be any privacy concerns, because the list acts within the scope if "configuring the source"
* _However_ I don't know off the top of my head if any change in this list will actually immediately impact the current program output. Which would mean that "just browsing" through the list of available windows would be immediately visible in the stream. And _that_ would have a privacy impact, however that impact would include the pure act of configuring the source itself.

My expectation for the window ID would be that `0` is an invalid ID, so the first element in the array should never contain a window with that ID at the first index.

Looking at the documentation for `NSWindow windowNumber` it states:

> If the window doesn’t have a window device, the value of this property is equal to or less than 0

The docs for `windowNumbersWithOptions` however state that it would only return the window numbers for _visible_ windows, which implies that they all have a window device (because otherwise they wouldn't be visible).

Device in this case means the `NSDevice` a window is attached to.

Reply:
> * Logically it should not be an issue to fall back on the very first item in the list, because you're configuring the source and you haven't committed the change yet - you're just setting it up. In that case there should not be any privacy concerns, because the list acts within the scope if "configuring the source"
> * _However_ I don't know off the top of my head if any change in this list will actually immediately impact the current program output. Which would mean that "just browsing" through the list of available windows would be immediately visible in the stream. And _that_ would have a privacy impact, however that impact would include the pure act of configuring the source itself.

I'm mainly thinking about the case where OBS has just launched and there are scenes containing window captures from a prior session. The current behavior is that these sources will hit this code on startup and initialize pointing to nil (this is what "usually" happens), or if there happens to be a window on the system with the same `SCWindow.windowID`, it captures that window. Both seem undesirable.

This change would cause all these sources in the latter case to capture whatever window happens to be the first in the list provided. To me this seems as undesirable as capturing a window that happens to have the same ID.

> Looking at the documentation for `NSWindow windowNumber` it states:

I'm not sure if this corresponds to SCWindow's windowID. I don't believe it does?

Reply:
That's a good catch, I didn't consider the use case of OBS launching and not being able to restore capture sessions for the old windows.

In that case the capture should indeed abort early like your PR suggests or - if we want to be fancy about it - show a placeholder.

Reply:
👍 I will do an updated push in a bit.